nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/AppLibClassLoaderServiceImpl.java

```diff
@@ -129,6 +140,55 @@ public DelegatingClassLoader run() {
         return (DelegatingClassLoader.ClassFinder)appLibClassFinder;
     }
 
+    private static class JarFileClassFinder extends ClassLoader implements DelegatingClassLoader.ClassFinder {
+
+        private final URI libUri;
+
+        JarFileClassFinder(URI libUri, ClassLoader parent) {
+            super(parent);
+            this.libUri = libUri;
+        }
+
+        @Override
+        public Class<?> findClass(String name) throws ClassNotFoundException {
+            try (JarFile jarFile = new JarFile(new File(libUri), false, ZipFile.OPEN_READ, Runtime.version())) {
+                String path = name.replace('.', '/').concat(".class");
+                JarEntry jarEntry = jarFile.getJarEntry(path);
+                if (jarEntry != null) {
```

Comment:
I would check it for null (inverted if) and throw CNFE, then you have one inner block level less.

Reply:
Another issue might be that going to the file system for every class, opening the file and unpacking the entry we want might not be effective ... but ... the original code does that too, am I right? And it probably has to ... because in theory there might be overlapping packages, resources, etc ... 

Btw do we really always want to return null for resources?
Also I am now not sure how it is with Java Modules, I am always forgetting these things ... we should start writing tests using these features.

Maybe add some logs so we could watch the sequence.

Reply:
> Another issue might be that going to the file system for every class, opening the file and unpacking the entry we want might not be effective

There will be overhead when opening and closing a file. But this will get rid of the problem of blocking the file at the filesystem level on the OSes with mandatory file locking.

> because in theory there might be overlapping packages, resources, etc ...



> Btw do we really always want to return null for resources?

This is a skeleton now. All methods will be implemented, also many other checks will be added. )

> Also I am now not sure how it is with Java Modules

ClassFinder`s doesnt't deal with Java Module System at this time. This ability should be implemented separately.